lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -1264,12 +1264,21 @@ def get_assigned_certificates(self) -> Tuple[List[ProviderCertificate], PrivateK
     def _find_certificate_in_relation_data(
         self, csr: RequirerCertificateRequest
     ) -> Optional[ProviderCertificate]:
-        """Return the certificate that match the given CSR."""
+        """Return the certificate that matches the given CSR, validated against the private key."""
+        if not self.private_key:
+            return None
         for provider_certificate in self.get_provider_certificates():
             if (
                 provider_certificate.certificate_signing_request == csr.certificate_signing_request
                 and provider_certificate.certificate.is_ca == csr.is_ca
             ):
+                if not provider_certificate.certificate_signing_request.matches_private_key(
+                    self.private_key
+                ):
+                    logger.warning(
+                        "Certificate does not match the private key. Ignoring invalid certificate."
+                    )
+                    continue
                 return provider_certificate
         return None
```

Comment:
The docstring says that we the certificate matches the CSR and the private key, but we do not really validate that the certificate returned matches the CSR or the private key. We check that the returned CSR is the same as the Requirer CSR, and that the CSR matches the private key. I think a better fix would be to check that the certificate itself matches the private key.

WDYT?

Reply:
I definitely agree, that was the intent - not sure how I got it wrong.

Reply:
This is fixed now